internal/scheduler/scheduler.go

```diff
@@ -61,7 +63,11 @@ func Scheduler(apps []argov1alpha1.Application, stage syncv1alpha1.ProgressiveSy
 		return scheduledApps
 	}
 
-	for i := 0; i < maxParallel-len(progressingApps); i++ {
+	// It may be that at point when apps were observed, no apps were progressing yet
+	// so maxParallel-len(progressingApps) might actually be greater than len(outOfSyncApps)
+	// causing the runtime to panic
```

Comment:
Do you have an example of this behaviour? Is there any test we can add so we can be sure we're solving that bug?

Reply:
It might be hard to write test for it, it happened to me whilst using a debugger because the debugger can pause all the execution while the rest of the world (ArgoCD) moves on. Basically what happened was:

```
maxParallel-len(progressingApps)
2
len(outOfSyncApps)
1
```

So there were no progressingApps at the moment, probably because they already got synced while the debugger was paused, but there was still a stale outOfSyncApps that was retrieved earlier on. I've repeated it several times in debugger, progressively stopping the world for less and less time, so it seems like this is a possible race condition that can happen.